### PR TITLE
Overwrite the printed caller with a correct one

### PR DESCRIPTION
**Pull request by vitaminmoo**
Add a VT100 escape sequencer to clear the line so the incorrect caller gets replaced, and have zap include the caller itself

### CODE REVIEW

File: logger.go [hunk 1]
```diff
@@ -72,12 +72,16 @@ func newLogger(w zapcore.WriteSyncer) *zap.Logger {
...
 }
 
 // Write logs all messages as logs via o.
 func (o testOutput) Write(p []byte) (int, error) {
-	msg := strings.TrimSpace(string(p))
+	// The escape sequence clears the line before writing the log message.
+	// This is unfortunately necessary as we can't call t.Helper() from all of
+	// zap's logging functions, which means t.Log will include an incorrect
+	// caller
+	msg := "\x1b[2K\r" + strings.TrimSpace(string(p))
```
Comment (fgrosse) [inline logger.go:84]: Hey Graham, thanks for the PR!

I get why you want to do this but I'm very hesitant to merge this because it might break compatibility on Windows and can lead to surprising results in some cases. For instances, tests can be executed with `-json`. The escape sequence will be visible there:

```
❯ go test -v -json | grep "This is a debug message" | jq
{
  "Time": "2024-11-27T10:41:05.668582371+01:00",
  "Action": "output",
  "Package": "github.com/fgrosse/zaptest",
  "Test": "TestLogger",
  "Output": "    logger.go:85: \u001b[2K\rDEBUG\tzaptest/logger_test.go:16\tThis is a debug message, debug messages will be logged as well\n"
}
```

I tried myself if I could find a good solution but I also don't see an alternative. The only thing we can do is to call `t.Helper()` in this function so the test output points to zap/core (i.e., `core.go:99`) but I'm not sure it really improves anything.

I checked what [`go.uber.org/zap/zaptest`](https://github.com/uber-go/zap/tree/master/zaptest) does and it seems they also just accept the prefix and don't do anything about it.

So, in summary, I don't see a good way to support this feature in a clean way with the current `testing.TB` Log function :shrug:

**Comment by fgrosse** [comment]
Closing stale PR
